Hard/MinimumWindowSubstring.java

```diff
@@ -0,0 +1,76 @@
+import java.util.HashMap;
+
+/**
+ * Minimum Window Substring
+ * Given a string S and a string T, find the minimum window in S which will contain all the characters in T in complexity O(n).
+ * For example,
+ * S = "ADOBECODEBANC"
+ * T = "ABC"
+ * Minimum window is "BANC".
+ * Tags: Hash Table Two Pointers String
+ * Similar Problems:  (H) Substring with Concatenation of All Words (M) Minimum Size Subarray Sum (H) Sliding Window Maximum
+ * 
+ * Analysis:
+ * 1.Store the char in string t as a key and times appearing as value into a hashmap;
+ * 2.If char at s is contained in hashmap, value(times) in map - 1 and count ++;
+ * 3.When count = t.length(), minLenght and minStart records the min number
+ * 4.Slide window and let the left point to the first char in s ,which is in map  
+ *        s                        t        
+ * A D O B E C O D E B A N C      A B C
+ * A D O B E C 
+ * B E C O D E B A
+ * C O D E B A
+ * B A N C
+ * 
+ * @author chenshuna
+ */
+
+public class MinimumWindowSubstring {
+    public static String minWindow(String s, String t) {
+        if(s.length() == 0 || t.length() ==0 || s.length() < t.length()) {
+            return "";
+        }
+        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
+        for(char c : t.toCharArray()) {
+            if(map.containsKey(c)) {
```

Comment:
can be more concise like `map.put(c, (map.containsKey(c) ? map.get(c) : 0) + 1);`